Update expected error messages for 3.11.0

jadams-tresys opened the pull request:
Due to changes made for DAFFODIL-2948, the mode of the error, eg "Validation Error" or "Parse Error" is not included in Diagnostic.getMessage() anymore.  It is instead contained in Diagnostic.getModeName().

stevedlawrence reviewed (approved):
+1

mbeckerle commented on hexWords/src/test/java/TestHexWords.java:
```diff
@@ -126,16 +126,14 @@ private void compare(String expectedXML, Document doc) throws IOException {
-        assertTrue(r.diags.stream().map(d -> d.getMessage() ).anyMatch( m ->
-            m.contains("Validation Error") &&
-                    m.contains("illegal")));
+        assertTrue(r.diags.stream().map(d -> d.getModeName() ).anyMatch( m -> m.contains("Validation Error")));
```
This makes the project dependent on Daffodil 3.11.0. 
Why not just call toString() instead of getMessage() and then use the original string matching for Validation Error and illegal. 
That makes the tests and this project portable to Daffodil versions prior to 3.11.0.

jadams-tresys replied:
Agreed, that makes more sense.  Should be handled in my most recent commit.

mbeckerle reviewed (changes requested):
Change to call toString instead of getMessage. toString outputs the original message with all the words in it. 
Avoids dependency on Daffodil 3.11.0.

Commit:
Check against Diagnostic.toString for errors

Due to changes made in DAFFODIL-2948, Diagnostic.getMessage no longer
includes the mode of the diagnostic, ie "Validation Error" or "Parse
Error". This data is still part of Diagnostic.toString, so use that
instead when checking the contents of error messages.

mbeckerle commented on hexWords/src/test/java/TestHexWords.java:
```diff
@@ -126,16 +126,12 @@ private void compare(String expectedXML, Document doc) throws IOException {
-        assertTrue(r.diags.stream().map(d -> d.getMessage() ).anyMatch( m ->
-                m.contains("Validation Error") &&
-                        m.contains("illegal")));
+        assertTrue(r.diags.stream().map(d -> d.toString() ).anyMatch( m -> m.contains("Validation Error") && m.contains("illegal")));
```
This is certainly the right way to fix this going forward as it is the least change to the code.

However, I have to say this gives me some pause as it is a backward incompatible change in the API, and I think this particular change should be reserved for Daffodil v4.0.0. Or rather, the new methods getMode or whatever can be added, the behavior of getMessage() changing should not be done until 4.0.0, if at all. We should consider defining a getMessageOnly() method or some such to have the new behavior and leave getMessage() alone. 

@stevedlawrence I believe I saw the PR of this API change from you (I could be wrong). Was this API change needed to address some specific user issue?

mbeckerle replied:
We might even add getMessageOnly() and deprecate getMessage() in this release. Then for v4.0.0 we can remove getMessage() or change its behavior.

stevedlawrence replied:
Yes, this was requested by the smooks.org devs: https://lists.apache.org/list?users@daffodil.apache.org:2025-3

Here's the PR: https://github.com/apache/daffodil/pull/1478

My thinking was that this is really the correct behavior for getMessage (I don't view the mode and schema locations as "the message") and it's a straightforward change for API users to switch to toString if they rely on the old behavior. And I would argue people shouldn't rely on getMessage returning a certain form--I don't really consider our diagnostic message format as something that is stable, just the API that gets messages. But I don't feel strongly about that and I see the other side of the argument.

Adding something like getMessageOnly is an easy enough change. Though I'm not sure we want to deprecate getMessage since it's part of the Exception API that Diagnostic implements.

stevedlawrence replied:
I've opened https://github.com/apache/daffodil/pull/1493 to revert to the old behavior for discussion

mbeckerle replied:
> My thinking was that this is really the correct behavior for getMessage (I don't view the mode and schema locations as "the message") and it's a straightforward change for API users to switch to toString if they rely on the old behavior. And I would argue people shouldn't rely on getMessage returning a certain form--I don't really consider our diagnostic message format as something that is stable, just the API that gets messages. But I don't feel strongly about that and I see the other side of the argument.

Ok. Thanks for reviewing that for me. I concur that this is an API improvement.

In theory updating a schema from 3.10.0 to 3.11.0 should only require a recompile. I guess we're weakening that to say "unless your schema depends on error message contents, which are subject to change even for minor releases." I'm ok with that - in fact I'm in favor of us having the flexibility to improve diagnostics in point releases, since they are a major class of usability bugs.

mbeckerle reviewed (approved):
+1

I want to have the discussion of whether the getMessage() change should be in 3.11.0 or not. 

But orthogonal to that this change set is fine. 

Note that because this calls toString instead of getMessage, this change is orthogonal to the issue of whether 3.11.0 or 4.0.0 changes the behavior of getMessage.
